rviz_common/src/rviz_common/depth_cloud_mld.cpp

```diff
@@ -0,0 +1,601 @@
+  std::vector<uint32_t> & rgba_color_raw,
+  ros_integration::RosNodeAbstractionIface::WeakPtr rviz_ros_node)
+{
+  int width = depth_msg->width;
+  int height = depth_msg->height;
+
+  sensor_msgs::msg::PointCloud2::SharedPtr cloud_msg = initPointCloud();
+  cloud_msg->data.resize(height * width * cloud_msg->point_step * 2);
+
+  uint32_t * color_img_ptr = nullptr;
+
+  if (!rgba_color_raw.empty()) {
+    color_img_ptr = &rgba_color_raw[0];
+  }
+
+  ////////////////////////////////////////////////
+  // depth map to point cloud conversion
+  ////////////////////////////////////////////////
+
+  float * cloud_data_ptr = reinterpret_cast<float *>(&cloud_msg->data[0]);
+  uint8_t * cloud_shadow_buffer_ptr = &shadow_buffer_[0];
+
+  const std::size_t point_step = cloud_msg->point_step;
+
+  std::size_t point_count = 0;
+  std::size_t point_idx = 0;
+
+  double time_now = rviz_ros_node.lock()->get_raw_node()->now().seconds();
+  double time_expire = time_now - shadow_time_out_;
+
+  const T * depth_img_ptr = reinterpret_cast<const T *>(&depth_msg->data[0]);
+
+  std::vector<float>::iterator proj_x;
+  std::vector<float>::const_iterator proj_x_end = projection_map_x_.end();
+
+  std::vector<float>::iterator proj_y;
+  std::vector<float>::const_iterator proj_y_end = projection_map_y_.end();
+
+  // iterate over projection matrix
+  for (proj_y = projection_map_y_.begin(); proj_y != proj_y_end; ++proj_y) {
+    for (proj_x = projection_map_x_.begin(); proj_x != proj_x_end;
+      ++proj_x, ++point_idx, ++depth_img_ptr, cloud_shadow_buffer_ptr += point_step)
+    {
+      // lookup shadow depth
+      float shadow_depth = shadow_depth_[point_idx];
+
+      // check for time-outs
+      if ((shadow_depth != 0.0f) && (shadow_timestamp_[point_idx] < time_expire)) {
+        // clear shadow pixel
+        shadow_depth = shadow_depth_[point_idx] = 0.0f;
+      }
+
+      T depth_raw = *depth_img_ptr;
+      if (DepthTraits<T>::valid(depth_raw)) {
+        float depth = DepthTraits<T>::toMeters(depth_raw);
+
+        // pointer to current point data
+        float * cloud_data_pixel_ptr = cloud_data_ptr;
+
+        // define point color
+        uint32_t color;
+        if (color_img_ptr) {
+          color = *color_img_ptr;
+        } else {
+          color = ((uint32_t)255 << 16 | (uint32_t)255 << 8 | (uint32_t)255);
+        }
+
+        // fill in X,Y,Z and color
+        *cloud_data_ptr = (*proj_x) * depth;
+        ++cloud_data_ptr;
+        *cloud_data_ptr = (*proj_y) * depth;
+        ++cloud_data_ptr;
+        *cloud_data_ptr = depth;
+        ++cloud_data_ptr;
+        *cloud_data_ptr = *reinterpret_cast<float *>(&color);
+        ++cloud_data_ptr;
+
+        ++point_count;
+
+        // if shadow point exists -> display it
+        if (depth < shadow_depth - shadow_distance_) {
+          // copy point data from shadow buffer to point cloud
+          memcpy(cloud_data_ptr, cloud_shadow_buffer_ptr, point_step);
+          cloud_data_ptr += 4;
```

Comment:
This code comes from RViz 1 and as far as I can understand, the `cloud_data_ptr` includes X, Y, Z and Color, that's why incrise the pointer in 4 positions. does it makes sense ?

Reply:
Let me make this slightly more concrete.

Let's assume that right before this `memcpy`, `cloud_data_ptr` points to address 0x20 (obviously not realistic, but just to make the math easy).  Let's further assume that point_step is 16 bytes (this follows from the `POINT_STEP` at the top of this file).  In this case, the `memcpy` will copy 16 bytes from `cloud_shadow_buffer_ptr` into `cloud_data_ptr`, and then the following line will increment the pointer by 16 bytes (4*sizeof(float)).  That all works out.

But let's assume that `point_step` is actually *not* 16 bytes, and is instead 12 bytes (I feel like that situation can happen, since the size of the point step is coming from the delivered message).  In that case, the `memcpy` will copy 12 bytes from `cloud_shadow_buffer_ptr` into `cloud_data_ptr`, and then the following line will increment the pointer by 16 bytes (4*sizeof(float)).  That means we skipped 4 bytes in there that we should not have.

Actually, this somewhat also calls into question the sizing of the `shadow_buffer_` itself in `initializeConversion`, since that is always hard-coded as `width*height*POINT_STEP`.  That feels like it could be wrong; should that be `width*height*point_step`?

Reply:
uhmmm, 

`initializeConversion` intialize the `shadow_buffer_` using de `depth_msg` which is a `sensor_msgs::msg::Image`, there is no corresponds with points yet. Up to this point we can't use the `point_step`. This is set to 16.

Then in the `generatePointCloudML` method loop, we fill the `cloud_shadow_buffer_ptr` data with in `X`,`Y`,`Z` and `color`. And this is always fixed: 4 float, this is why the `shadow_buffer_` is always setted to this size (width*height*POINT_STEP).

Then in `generatePointCloudML` the `cloud_msg` is initialize in the `initPointCloud` which set the point_step to `16`, no matter what. Which means that `point_step` is always 16 (== 4 * size(float)).

Does it make sense ?

Reply:
Ah, I see.  You are right.  I thought that `cloud_msg` was coming in from the user, but as you mention it is actually being generated internally via `initPointCloud`.

So in that case, I think this is safe.  Thanks for the explanation.